pubtools/pulplib/_impl/client/repo_lock.py

```diff
@@ -0,0 +1,161 @@
+import json
+import random
+import string
+from time import sleep
+
+from datetime import datetime, timedelta
+from more_executors.futures import f_map
+
+
+LOCK_CLAIM_STR = "lock-claim-"
+
+LOCK_VALID_FROM_OFFSET = 10
+LOCK_EXPIRATION_OFFSET = 3600
+
+# How long to sleep before checking if we can activate the lock again
+# This should be a sleep per locks ahead of us. If there's 3 locks ahead of
+# our one, it's probably better to wait longer before checking again.
+AWAIT_LOCK_ACTIVATION_SLEEP = 15
+
+REPO_ENDPOINT_BASE = "pulp/api/v2/repositories/%s/"
+
+
+# Helper class for holding data and convenience functions.
+class LockClaim(object):
+    def __init__(self, **kwargs):
+        self.id = (
+            kwargs["id"]
+            if "id" in kwargs
+            else "".join(random.choices(string.ascii_letters + string.digits, k=8))
+        )
+        self.created = (
+            datetime.fromtimestamp(kwargs["created"])
+            if "created" in kwargs
+            else datetime.now()
+        )
+        self.valid_from = (
+            datetime.fromtimestamp(kwargs["valid_from"])
+            if "valid_from" in kwargs
+            else self.created + timedelta(seconds=LOCK_VALID_FROM_OFFSET)
+        )
+        self.expires = (
+            datetime.fromtimestamp(kwargs["expires"])
+            if "expires" in kwargs
+            else self.created + timedelta(seconds=LOCK_EXPIRATION_OFFSET)
+        )
+
+    @property
+    def as_json(self):
+        data = {
+            "id": self.id,
+            "created": datetime.timestamp(self.created),
+            "valid_from": datetime.timestamp(self.valid_from),
+            "expires": datetime.timestamp(self.expires),
+        }
+        return json.dumps(data)
+
+    def __lt__(self, other):
+        return self.created < other.created
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    @staticmethod
+    def from_json_data(json_data):
+        return LockClaim(**json_data)
+
+    @property
+    def is_valid(self):
+        return self.expires > datetime.now()
+
+
+class RepoLock(object):
+    def __init__(self, repo_id, request_func):
+        self._repo_id = repo_id
+        self._request_func = request_func
+        self._lock_claim = None
+
+    def __enter__(self):
+        self.remove_expired_locks()
+        self._lock_claim = LockClaim()
+        self.submit_lock_claim()
+        self.await_lock_activation()
+
+    def __exit__(self, exc_type, exc_value, tb):
+        self.delete_lock_claim()
+        self._lock_claim = None
+
+    def get_repo_lock_data(self, filter_invalid=False):
+        print("Getting current repo lock data")
```

Comment:
When you're done with these, rather than removing the prints you should probably change them to debug or info logs.